llvm/include/llvm/ADT/SmallVector.h

```diff
@@ -100,7 +100,18 @@ template <class Size_T> class SmallVectorBase {
   /// This does not construct or destroy any elements in the vector.
   void set_size(size_t N) {
     assert(N <= capacity());
-    Size = N;
+    Size = static_cast<Size_T>(N);
+    assert(size() == N);
+  }
+
+  /// Set the array capacity to \p M, which the current array size must
+  /// not be greater than.
+  ///
+  /// This does not construct or destroy any elements in the vector.
+  void set_capacity(size_t M) {
+    assert(size() <= M);
```

Comment:
I think this assertion is not useful, and may be misleading.

An assertion that would be clear and useful here would be:
```
assert(M <= SizeTypeMax());
```
...before assigning the elements and capacity.

You also wouldn't need `assert(capacity() == M);` as this assertion implies it will be successful. So remove that.

Reply:
thanks, the assertion you propose would indeed be useful!

`size() <= M` is to just guarantee that we don't end up in a weird container state where `Size > Capacity`. but then using set_size() together with set_capacity() in the current sense would be problematic:

```
// capacity = 50, size = 40

// grow:
// set_size(70); // asserts: 70 <= capacity() fails
set_capacity(100); // OK
set_size(70); // OK

// shrink:
// set_capacity(10); // asserts: size() <= 10 fails
set_size(2); // OK
set_capacity(10); // OK
```

I guess due to this "swap" in the order, it might make it harder to write generic code? but then I again think of `Size > Capacity` problem.

Reply:
No, if the existing `assert(N <= capacity());` passes then I think `assert(M <= SizeTypeMax());` must be true.

I think:
* `set_size` should only check `assert(N <= capacity());`
* `set_capacity` should only check `assert(M <= SizeTypeMax());`

Using set_capacity to shrink (regardless of the size) doesn't make sense, because `set_capacity` would not actually handle shrinking (only a small piece of shrinking - switching to a smaller allocation).

Reply:
done